### DISCUSSION TITLE
Render markdown sections separately

### DISCUSSION THREAD

**Discussion by virvum** [yuin/goldmark#186] [Q&A]
I would like to split a markdown document into sections, separated by headlines (preserving the headline in each section). I'd also be interested in rendering a single section only.

A quick and dirty solution would be to render the entire document and then parse the HTML output using golang.org/x/net/html or something similar and split up the output, but I'd prefer to do this using a custom render procedure.

A quick example of what I'd like to achieve:

Input:
```
# Heading 1
Text below heading 1
## Heading 2
Text below heading 2
## Heading 3
Text below heading 3
```

Output:
```
[]string{
    "<h1>Heading 1</h1>\n<p>Text below heading 1</p>",
    "<h2>Heading 2</h2>\n<p>Text below heading 2</p>",
    "<h2>Heading 3</h2>\n<p>Text below heading 3</p>",
}
```

Any hints are welcome!

**Comment by yuin** [comment]
You can use `ast.Walk` to find heading elements. 
Then you can render these headings by `renderer.Render`.

**Comment by nerg4l** [comment]
You could do the following with `ast.Walk`:

```go
	md := goldmark.New()
	reader := text.NewReader(markdownContent)
	doc := md.Parser().Parse(reader)
	// AppendChild removes the node from doc. To be able to use Walk correctly
	// we shouldn't modify the original doc in Walk.
	var segments [][]ast.Node
	ast.Walk(doc, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
		if !entering || n.Type() == ast.TypeDocument {
			return ast.WalkContinue, nil
		}
		if _, ok := n.(*ast.Heading); ok {
			segments = append(segments, nil)
		}
		if len(segments) > 0 {
			segment := segments[len(segments)-1]
			segment = append(segment, n)
			segments[len(segments)-1] = segment
		}
		return ast.WalkSkipChildren, nil
	})

	var sections []string
	for i := range segments {
		doc := ast.NewDocument()
		for _, node := range segments[i] {
			doc.AppendChild(doc, node)
		}
		var b strings.Builder
		if err := md.Renderer().Render(&b, markdownContent, doc); err != nil {
			t.Fatal(err)
		}
		sections = append(sections, b.String())
	}
	// Do what you need to do
	log.Printf("%#v", sections)
```

**Comment by nnsite** [comment]
I use hugo.
What kind of solution could be used to get a layout like this one:
```
<section><h1>title</h1><p1>Preamble
<section><h2>A heading n°1</h2><p>text</section>
<section><h2>B heading n°1</h2><p>text<section>
```
It's a very common layout, advised by the w3c. Goldmark should make it possible easily.
I think there should be options flags or options for the most common layouts. The normal without section, this one and the one with the heading number being reset within each section layer.
